src/engine/Service.cpp

```diff
@@ -140,60 +143,76 @@ ProtoResult Service::computeResult([[maybe_unused]] bool requestLaziness) {
+      }
 
-    if (jsonResult.empty()) {
-      throwErrorWithContext("Response from SPARQL endpoint is empty");
+      resVariables = jsonResult["head"]["vars"].get<std::vector<std::string>>();
+      resBindings =
+          jsonResult["results"]["bindings"].get<std::vector<nlohmann::json>>();
+    } catch (const nlohmann::json::parse_error&) {
+      throwErrorWithContext("Failed to parse the SERVICE result as JSON");
+    } catch (const nlohmann::json::type_error&) {
+      throwErrorWithContext("JSON result does not have the expected structure");
     }
 
-    resVariables = jsonResult["head"]["vars"].get<std::vector<std::string>>();
-    resBindings =
-        jsonResult["results"]["bindings"].get<std::vector<nlohmann::json>>();
-  } catch (const nlohmann::json::parse_error&) {
-    throwErrorWithContext("Failed to parse the SERVICE result as JSON");
-  } catch (const nlohmann::json::type_error&) {
-    throwErrorWithContext("JSON result does not have the expected structure");
-  }
+    // Check if result header row is expected.
+    std::string headerRow =
+        absl::StrCat("?", absl::StrJoin(resVariables, " ?"));
+    std::string expectedHeaderRow = absl::StrJoin(
+        parsedServiceClause_.visibleVariables_, " ", Variable::AbslFormatter);
+    if (headerRow != expectedHeaderRow) {
+      throwErrorWithContext(absl::StrCat(
+          "Header row of JSON result for SERVICE query is \"", headerRow,
+          "\", but expected \"", expectedHeaderRow, "\""));
+    }
 
-  // Check if result header row is expected.
-  std::string headerRow = absl::StrCat("?", absl::StrJoin(resVariables, " ?"));
-  std::string expectedHeaderRow = absl::StrJoin(
-      parsedServiceClause_.visibleVariables_, " ", Variable::AbslFormatter);
-  if (headerRow != expectedHeaderRow) {
-    throwErrorWithContext(absl::StrCat(
-        "Header row of JSON result for SERVICE query is \"", headerRow,
-        "\", but expected \"", expectedHeaderRow, "\""));
+    // Set basic properties of the result table.
+    IdTable idTable{getExecutionContext()->getAllocator()};
+    idTable.setNumColumns(getResultWidth());
+    LocalVocab localVocab{};
+    // Fill the result table using the `writeJsonResult` method below.
+    size_t resWidth = getResultWidth();
+    CALL_FIXED_SIZE(resWidth, &Service::writeJsonResult, this, resVariables,
+                    resBindings, &idTable, &localVocab);
+
+    return {std::move(idTable), resultSortedOn(), std::move(localVocab)};
+  } catch (const std::runtime_error& e) {
+    // if the `SILENT` keyword is set in the service clause, catch the error and
+    // return a neutral IdTable
+    if (parsedServiceClause_.silent_) {
+      IdTable idTable{getExecutionContext()->getAllocator()};
+      idTable.setNumColumns(getResultWidth());
+      idTable.emplace_back();
+      for (size_t colIdx = 0; colIdx < getResultWidth(); ++colIdx) {
+        idTable(0, colIdx) = Id::makeUndefined();
+      }
```

Comment:
You can even factor this setting up of the IdTable into a small helper function
`makeNeutralElementResultForSilentFail` or something.